stdlib/array.ml

```diff
@@ -442,8 +442,9 @@ let fast_sort = stable_sort
 let shuffle ~rand a = (* Fisher-Yates *)
   for i = length a - 1 downto 1 do
     let j = rand (i + 1) in
+    if not (0 <= j && j <= i) then invalid_arg "Array.shuffle";
```

Comment:
I would want a better error message than this, for users to know what is going on. For example `"Array.shuffle: `rand` returned an integer out of range"`. Or one could even be more clear:

```ocaml
Printf.ksprintf failwith
  "Array.shuffle: `rand` returned an integer %d \
   out of the expected range [0; %d]"
  j i
```